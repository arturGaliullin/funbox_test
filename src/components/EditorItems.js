import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class EditorItems extends Component {
    onDelete = (e, id) => {
        e.preventDefault();
        this.props.removePlacemark(id);
    };

    onDragStart = (e, i) => {
        this.draggedItem = this.props.placemarks[i];
        e.dataTransfer.effectAllowed = 'move';
        e.dataTransfer.setData('text/html', e.target);
        e.dataTransfer.setDragImage(e.target, 20, 20);
    };

    onDragOver = i => {
        const draggedOverItem = this.props.placemarks[i];

        if (this.draggedItem === draggedOverItem) {
            return;
        }

        const items = this.props.placemarks.filter(placemark => placemark.id !== this.draggedItem.id);

        items.splice(i, 0, this.draggedItem);

        this.props.updatePlacemarks(items);
    };

    onDragEnd = () => {
        this.draggedIdx = null;
    };

    render() {
        return (
            <div className='editor__items'>
                {this.props.placemarks.map((placemark, i) =>
                    <div key={placemark.id} className='editor__item' onDragOver={() => this.onDragOver(i)}>
                        <div
                            draggable
                            onDragStart={(e) => this.onDragStart(e, i)}
                            onDragEnd={this.onDragEnd}
                        >
                            <span className='editor__item-name'>{placemark.name}</span>
                            <button onClick={(e) => this.onDelete(e, placemark.id)} className="button editor__item-button">X</button>
                        </div>
                    </div>

                )}
            </div>
        )
    }
}

EditorItems.propTypes = {
    removePlacemark: PropTypes.func,
    updatePlacemarks: PropTypes.func,
    placemarks: PropTypes.array
};