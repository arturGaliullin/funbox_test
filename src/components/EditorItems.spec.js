import React from 'react';
import { shallow, mount } from 'enzyme';
import EditorItems from './EditorItems';

const props = {
    placemarks: [],
    removePlacemark: () => {},
    updatePlacemarks: () => {},
};

describe('EditorItems', () => {
    it('should render with props', () => {
        const component = shallow(<EditorItems {...props} />);

        expect(component).toMatchSnapshot();
    });

    it('test delete placemark', () => {
        props.placemarks = [{id: 1, name: 'name'}, {id: 2, name: 'name-1'}];
        const component = shallow(<EditorItems {...props} />);
        const buttonClose = component.find('.editor__item-button').first();
        buttonClose.simulate('click', {preventDefault: () => {}});
        component.instance().removePlacemark = jest.fn();
        component.update();
        component.instance().removePlacemark({preventDefault: () => {}}, props.placemarks[0].id);
        expect(component.instance().removePlacemark).toHaveBeenCalled();
    });
});