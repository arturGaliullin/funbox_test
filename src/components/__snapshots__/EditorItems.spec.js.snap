// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorItems should render with props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditorItems
    placemarks={Array []}
    removePlacemark={[Function]}
    updatePlacemarks={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [],
      "className": "editor__items",
    },
    "ref": null,
    "rendered": Array [],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [],
        "className": "editor__items",
      },
      "ref": null,
      "rendered": Array [],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
